pydantic/typing.py

```diff
@@ -363,6 +368,38 @@ def is_typeddict(type_: Type[Any]) -> bool:
     return lenient_issubclass(type_, dict) and hasattr(type_, '__total__')
 
 
+def is_typeddict_special(type_: Type[Any]) -> bool:
+    """
+    Check if type is a TypedDict special form (Required or NotRequired).
+    """
+    return type_ is TypedDictRequired or type_ is TypedDictNotRequired
+
+
+def update_typeddict_keys(
+    typeddict_cls: Type['TypedDict'],  # type: ignore[valid-type]
+) -> None:
```

Comment:
how come this is required if `typeddict_cls.__required_keys__` and `typeddict_cls.__optional_keys__` are already set?

Might be worth adding an explanation to the docstring.

Reply:
Please, review new doc string.

Reply:
I'm not really for it. The issue comes from `typing.TypedDict` in 3.10, which doesn't support that.
But `TypedDict` from `typing` in 3.11 and `typing_extensions` supports it.
Couldn't we just warn the user to not use `typing.TypedDict` if the version is `3.10` and `Required` or `NotRequired` is used?